### PR TITLE
Domain fix

### LINKED ISSUE #3284: Invalid data source name in output scale domain for faceted layer
Here is a minimal spec to reproduce the problem. 

```
{
  "data": {"url": "data/stocks.csv"},
  "facet": {
    "row": {"field": "symbol", "type": "nominal"}
  },
  "spec": {
    "layer": [{
      "data": {"values": [{"ref": 100}]},
      "mark": "rule",
      "encoding": {
        "y": { "field":"ref", "type": "quantitative"},
        "size": {"value": 2}
      }
    }] 
  }
}
```

This produces the following error: 

```
Error: Undefined data set name: "scale_child_layer_0_main"
```

Originally reported by @eibanez in https://github.com/vega/vega-lite/issues/1601#issuecomment-358978898

### PR DESCRIPTION
**Pull request by domoritz**
Fixes #3284

### CODE REVIEW

File: src/compile/model.ts [hunk 1]
```diff
@@ -186,7 +186,9 @@ export abstract class Model {
         sources: parent ? parent.component.data.sources : {},
         outputNodes: parent ? parent.component.data.outputNodes : {},
         outputNodeRefCounts: parent ? parent.component.data.outputNodeRefCounts : {},
-        ancestorParse: parent ? {...parent.component.data.ancestorParse} : {}
+        ancestorParse: parent ? {...parent.component.data.ancestorParse} : {},
+        // data is faceted if the spec is a facet spec or the parent has faceted data and no data is defined
+        isFaceted: isFacetSpec(spec) || (parent && parent.component.data.isFaceted && !spec.data)
```
Comment (kanitw) [inline src/compile/model.ts]: If I understand correctly, this will also be correct for facet of layer right?  (Just asking to make sure)

**Comment by kanitw** [comment]
Test failing

File: src/compile/model.ts (hunk 1 above)
Comment (domoritz) [reply src/compile/model.ts]: Yeah, it works with my example.

**Comment by domoritz** [comment]
Revised and added some more tests.

**Pull request merged**